buildbot_nix/github_projects.py

```diff
@@ -1,93 +1,229 @@
+    slugify_project_name,
+)
+from .projects import GitBackend, GitProject
+from .secrets import read_secret_file
 
-class GithubError(Exception):
-    pass
-
-
-class HttpResponse:
-    def __init__(self, raw: http.client.HTTPResponse) -> None:
-        self.raw = raw
-
-    def json(self) -> Any:
-        return json.load(self.raw)
-
-    def headers(self) -> http.client.HTTPMessage:
-        return self.raw.headers
-
-
-def http_request(
-    url: str,
-    method: str = "GET",
-    headers: dict[str, str] | None = None,
-    data: dict[str, Any] | None = None,
-) -> HttpResponse:
-    body = None
-    if data:
-        body = json.dumps(data).encode("ascii")
-    if headers is None:
-        headers = {}
-    headers = headers.copy()
-    headers["User-Agent"] = "buildbot-nix"
-
-    if not url.startswith("https:"):
-        msg = "url must be https: {url}"
-        raise GithubError(msg)
-
-    req = urllib.request.Request(  # noqa: S310
-        url, headers=headers, method=method, data=body
-    )
-    try:
-        resp = urllib.request.urlopen(req)  # noqa: S310
-    except urllib.request.HTTPError as e:
-        resp_body = ""
-        with contextlib.suppress(OSError, UnicodeDecodeError):
-            resp_body = e.fp.read().decode("utf-8", "replace")
-        msg = f"Request for {method} {url} failed with {e.code} {e.reason}: {resp_body}"
-        raise GithubError(msg) from e
-    return HttpResponse(resp)
-
-
-def paginated_github_request(url: str, token: str) -> list[dict[str, Any]]:
-    next_url: str | None = url
-    items = []
-    while next_url:
-        try:
-            res = http_request(
-                next_url,
-                headers={"Authorization": f"Bearer {token}"},
+
+class ReloadGithubProjects(BuildStep):
+    name = "reload_github_projects"
+
+    def __init__(self, token: str, project_cache_file: Path, **kwargs: Any) -> None:
+        self.token = token
+        self.project_cache_file = project_cache_file
+        super().__init__(**kwargs)
+
+    def reload_projects(self) -> None:
+        refresh_projects(self.token, self.project_cache_file)
+
+    @defer.inlineCallbacks
+    def run(self) -> Generator[Any, object, Any]:
+        d = threads.deferToThread(self.reload_projects)  # type: ignore[no-untyped-call]
+
+        self.error_msg = ""
+
+        def error_cb(failure: Failure) -> int:
+            self.error_msg += failure.getTraceback()
+            return util.FAILURE
+
+        d.addCallbacks(lambda _: util.SUCCESS, error_cb)
+        res = yield d
+        if res == util.SUCCESS:
+            # reload the buildbot config
+            os.kill(os.getpid(), signal.SIGHUP)
+            return util.SUCCESS
+        else:
+            yield self.addLog("log").addStderr(
+                f"Failed to reload project list: {self.error_msg}"
+            )
+            return util.FAILURE
+
+
+@dataclass
+class GithubConfig:
+    oauth_id: str
+
+    # TODO unused
+    buildbot_user: str
```

Comment:
Does `TODO unused` mean `buildbot_user` can be removed?

Reply:
At the time I wasn't a 100% sure, but now I am, and yes it can

Reply:
Thanks, I'll send a PR.